url8c14fd53ee_luca89_itaco_tgz-pJ8-main_RunnerJ8/src/compilatore/JasminTarget.java

```diff
@@ -0,0 +1,636 @@
+			String replacePattern, String replaceString) throws IOException {
+		/*
+		 * Copio il contenuto dello stub nel file di output
+		 */
+		InputStreamReader stub = new InputStreamReader(JasminTarget.class.getResourceAsStream("/compilatore/stub/"+stubFileName));
+		StringBuilder sb = new StringBuilder();
+		int charsRead = 0;
+		char[] fileData = new char[1024];
+
+		while (stub.ready()) {
+			charsRead = stub.read(fileData);
+			sb.append(fileData, 0, charsRead);
+		}
+		String stubReplaced = sb.toString().replaceAll(replacePattern,
+				replaceString);
+		outputFile.print(stubReplaced);
+	}
+
+	/*
+	 * Scrive nel linguaggio target delle parti di codice indipendenti dallo
+	 * specifico programma
+	 */
+	@SuppressWarnings("unused")
+	private void scriviContenutoStub(String stubFileName) throws IOException {
+		/*
+		 * Copio il contenuto dello stub nel file di output
+		 */
+		InputStream stub = JasminTarget.class.getResourceAsStream("/compilatore/stub/"+stubFileName);
+		int copiedBytes = 0;
+		int bytesRead = 0;
+		byte[] fileData = new byte[1024];
+
+		while (stub.available() > 0) {
+			bytesRead = stub.read(fileData);
+			outputFile.write(fileData, 0, bytesRead);
+			copiedBytes += bytesRead;
+		}
+	}
+
+	@Override
+	public void somma(Espressione addendo1, Espressione addendo2)
+			throws EccezioneSemantica {
+		addendo1.scriviCodice(this);
+		addendo2.scriviCodice(this);
+		bufferOutput.println("iadd");
+	}
+
+	@Override
+	public void caricaVariabile(String nome) throws EccezioneSemantica {
+		int id = idVariabile(nome);
+		bufferOutput.println("iload " + id);
+	}
+
+	@Override
+	public void costante(int costante) throws EccezioneSemantica {
+		bufferOutput.println("ldc " + costante);
+	}
+
+	@Override
+	public void sottrazione(Espressione minuendo, Espressione sottraendo)
+			throws EccezioneSemantica {
+		minuendo.scriviCodice(this);
+		sottraendo.scriviCodice(this);
+		bufferOutput.println("isub");
+	}
+
+	@Override
+	public void prodotto(Espressione fattore1, Espressione fattore2)
+			throws EccezioneSemantica {
+		fattore1.scriviCodice(this);
+		fattore2.scriviCodice(this);
+		bufferOutput.println("imul");
+	}
+
+	@Override
+	public void divisione(Espressione dividendo, Espressione divisore)
+			throws EccezioneSemantica {
+		dividendo.scriviCodice(this);
+		divisore.scriviCodice(this);
+		bufferOutput.println("idiv");
+	}
+
+	@Override
+	public void maggiore(Espressione parteSinistra, Espressione parteDestra)
+			throws EccezioneSemantica {
+		sottrazione(parteSinistra, parteDestra);
+		String labelMaggiore1 = generaLabel();
+		String labelMaggiore2 = generaLabel();
+		String ifgt = "ifgt ";
+		ifgt = ifgt + labelMaggiore1;
+		bufferOutput.println(ifgt);
+		bufferOutput.println("ldc -1");
+		String go_to = "goto ";
+		go_to = go_to + labelMaggiore2;
+		bufferOutput.println(go_to);
+		bufferOutput.println(labelMaggiore1 + ":");
+		bufferOutput.println("ldc 1");
+		bufferOutput.println(labelMaggiore2 + ": ");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 220:232, 244:256.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.